extensions/optaplanner/deployment/src/main/java/io/quarkus/optaplanner/deployment/OptaPlannerProcessor.java

```diff
@@ -0,0 +1,250 @@
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.RuntimeInitializedClassBuildItem;
+import io.quarkus.deployment.recording.RecorderContext;
+import io.quarkus.optaplanner.OptaPlannerBeanProvider;
+import io.quarkus.optaplanner.OptaPlannerObjectMapperCustomizer;
+import io.quarkus.optaplanner.OptaPlannerRecorder;
+
+class OptaPlannerProcessor {
+
+    OptaPlannerQuarkusConfig optaPlannerQuarkusConfig;
+
+    @BuildStep
+    FeatureBuildItem feature() {
+        return new FeatureBuildItem(FeatureBuildItem.OPTAPLANNER);
+    }
+
+    @BuildStep
+    void registerAdditionalBeans(BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        // The bean encapsulating the SolverFactory
+        additionalBeans.produce(new AdditionalBeanBuildItem(OptaPlannerBeanProvider.class));
+    }
+
+    @BuildStep(loadsApplicationClasses = true)
+    @Record(STATIC_INIT)
+    void recordSolverFactory(OptaPlannerRecorder recorder, RecorderContext recorderContext,
+            CombinedIndexBuildItem combinedIndex,
+            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchyClass,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<BeanContainerListenerBuildItem> beanContainerListener) {
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        SolverConfig solverConfig;
+        if (optaPlannerQuarkusConfig.solverConfigXml.isPresent()) {
+            String solverConfigXML = optaPlannerQuarkusConfig.solverConfigXml.get();
+            if (classLoader.getResource(solverConfigXML) == null) {
+                throw new IllegalStateException("Invalid optaplanner.solverConfigXML property (" + solverConfigXML
+                        + "): that classpath resource does not exist.");
+            }
+            solverConfig = SolverConfig.createFromXmlResource(solverConfigXML, classLoader);
+        } else if (classLoader.getResource(OptaPlannerQuarkusConfig.DEFAULT_SOLVER_CONFIG_URL) != null) {
+            solverConfig = SolverConfig.createFromXmlResource(
+                    OptaPlannerQuarkusConfig.DEFAULT_SOLVER_CONFIG_URL, classLoader);
+        } else {
+            solverConfig = new SolverConfig(classLoader);
+        }
+        // The deployment classLoader must not escape to runtime
+        solverConfig.setClassLoader(null);
+
+        IndexView indexView = combinedIndex.getIndex();
+        applySolverProperties(recorder, recorderContext, indexView, solverConfig);
+
+        if (solverConfig.getSolutionClass() != null) {
+            Type jandexType = Type.create(DotName.createSimple(solverConfig.getSolutionClass().getName()), Type.Kind.CLASS);
+            reflectiveHierarchyClass.produce(new ReflectiveHierarchyBuildItem(jandexType));
+        }
+        List<Class<?>> reflectiveClassList = new ArrayList<>();
```

Comment:
Maybe add a capacity, it looks like 5 is the max this can ever get, right?

Reply:
Fixed with new commit on this fork.